exonum/src/api_ng/node/public/explorer.rs

```diff
@@ -0,0 +1,166 @@
+    pub block: Block,
+    /// Precommits authorizing the block.
+    pub precommits: Vec<Precommit>,
+    /// Hashes of transactions in the block.
+    pub txs: Vec<Hash>,
+}
+
+/// Blocks in range parameters.
+#[derive(Debug, Serialize, Deserialize, PartialEq)]
+pub struct BlocksQuery {
+    /// The number of blocks to return. Should not be greater than `MAX_BLOCKS_PER_REQUEST`.
+    pub count: usize,
+    /// The maximum height of the returned blocks. The blocks are returned in reverse order,
+    /// starting from the latest and at least up to the `latest` - `count` + 1.
+    /// The default value is the height of the latest block in the blockchain.
+    pub latest: Option<Height>,
+    /// If true, then only non-empty blocks are returned. The default value is false.
+    #[serde(default)]
+    pub skip_empty_blocks: bool,
+}
+
+/// Block query parameters.
+#[derive(Debug, Serialize, Deserialize, PartialEq)]
+pub struct BlockQuery {
+    /// The height of the desired block.
+    pub height: Height,
+}
+
+/// Transaction query parameters.
+#[derive(Debug, Serialize, Deserialize, PartialEq)]
+pub struct TransactionQuery {
+    /// The hash of the transaction to be searched.
+    pub hash: Hash,
+}
+
+/// Exonum blockchain explorer API.
+pub trait ExplorerApi {
+    /// Corresponding error type for the API implementation.
+    type Error;
+
+    /// Returns the explored range and the corresponding headers. The range specifies the smallest
+    /// and largest heights traversed to collect at most count blocks.
+    fn blocks(&self, query: BlocksQuery) -> Result<BlocksRange, Self::Error>;
+
+    /// Searches for a transaction, either committed or uncommitted, by the hash.
+    fn transaction_info(
+        &self,
+        hash: TransactionQuery,
+    ) -> Result<Option<TransactionInfo>, Self::Error>;
+
+    /// Returns the content for a block of a specific height.
+    fn block(&self, query: BlockQuery) -> Result<Option<BlockInfo>, Self::Error>;
+
+    /// Adds explorer API endpoints to the corresponding scope.
+    fn wire(api_scope: &mut ServiceApiScope) -> &mut ServiceApiScope {
+        api_scope
+    }
+}
+
+impl ExplorerApi for ServiceApiState {
+    type Error = ApiError;
+
+    fn blocks(&self, query: BlocksQuery) -> Result<BlocksRange, Self::Error> {
+        let explorer = BlockchainExplorer::new(self.blockchain());
+        if query.count > MAX_BLOCKS_PER_REQUEST {
+            return Err(ApiError::BadRequest(format!(
+                "Max block count per request exceeded ({})",
+                MAX_BLOCKS_PER_REQUEST
+            )));
+        }
+
+        let (upper, blocks_iter) = if let Some(upper) = query.latest {
+            (upper, explorer.blocks(..upper.next()))
+        } else {
+            (explorer.height(), explorer.blocks(..))
+        };
+
+        let blocks: Vec<_> = blocks_iter
+            .rev()
+            .filter(|block| !query.skip_empty_blocks || !block.is_empty())
```

Comment:
I've spent several minutes trying to get why do we have 2 lambdas in the row :sweat: 
Can we somehow make this line less confusing?